src/mesh/domain.hpp

```diff
@@ -0,0 +1,183 @@
+//========================================================================================
+// Athena++ astrophysical MHD code
+// Copyright(C) 2014 James M. Stone <jmstone@princeton.edu> and other code contributors
+// Licensed under the 3-clause BSD License, see LICENSE file for details
+//========================================================================================
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#ifndef MESH_DOMAIN_HPP_
+#define MESH_DOMAIN_HPP_
+
+#include <array>
+#include <type_traits>
+#include <vector>
+
+#include "athena.hpp"
+
+namespace parthenon {
+
+struct IndexRange {
+  IndexRange() {}
+  IndexRange(int start, int end) : s(start), e(end) { assert(e >= s); }
+  int s = 0; /// Starting Index (inclusive)
+  int e = 0; /// Ending Index (inclusive)
+  int ncells() const noexcept { return e - s + 1; }
+};
+
+// Assuming we have a block
+//
+//  - - - - - - - - - -   ^
+//  |  |  ghost    |  |   |
+//  - - - - - - - - - -   |
+//  |  |     ^     |  |   |
+//  |  |     |     |  |
+//  |  | interior  |  |   entire
+//  |  |     |     |  |
+//  |  |     v     |  |   |
+//  - - - - - - - - - -   |
+//  |  |           |  |   |
+//  - - - - - - - - - -   v
+//
+enum class IndexDomain { entire, interior };
+
+//! \class IndexVolume
+//  \brief Defines the dimensions of a shape of indices
+//
+//  Defines the range of each dimension of the indices by defining a starting and stopping
+//  index also contains a label for defining which region the index shape is assigned too
+class IndexShape {
+ private:
+  std::array<IndexRange, NDIM> x_;
+  std::array<int, NDIM> entire_ncells_;
+
+  inline bool DimensionProvided_(const std::vector<int> &interior_dims, int dim) {
+    return dim <= interior_dims.size();
+  }
+
+  inline void MakeZeroDimensional_(int index) {
+    x_[index] = IndexRange(0, 0);
+    entire_ncells_[index] = 1;
+  }
+
+ public:
+  IndexShape() {}
+
+  IndexShape(const int &nx1, const int &nx2, const int &nx3, const int &ng)
```

Comment:
This is going to sound pedantic... but ti's convention throughout the code to order these nxs backwards: `nx3, nx2, nx1`. This has to do with which index moves fastest the array and the order of the loops. Typically it goes 3->2->1 in loop orderings.

Reply:
Fair point.